src/permission/rule-application.js

```diff
@@ -0,0 +1,420 @@
+var OPEN = 'open';
+var UNDEFINED = 'undefined';
+var LOADING = 'loading';
+var ERROR = 'error';
+var STRING = 'string';
+var EOL = require( 'os' ).EOL;
+
+var C = require( '../constants/constants' );
+var RecordRequest = require( '../record/record-request' );
+var messageParser = require( '../message/message-parser' );
+var JsonPath = require( '../record/json-path' );
+var utils = require( '../utils/utils' );
+
+/**
+ * This class handles the evaluation of a single rule. It creates
+ * the required variables, injects them into the rule function and
+ * runs the function recoursively until either all cross-references,
```

Comment:
runs the function recursively until either all cross-references and references to old or new data are loaded, when an error occurs or if it exceeds the maximum maxIterationCount